### ISSUE TITLE
Some allocation records disappear in between two consecutive dumps with -alloc_space or -alloc_objects

### ISSUE THREAD

**Issue by sangelovic** [gperftools/gperftools#1450]
Hello, we see the following strange behavior with gperftools v2.9.1 as well as latest v2.13.

In a dump `0001.heap`, processed with `-alloc_space` (and `-top`) pprof options, we get the following output:

```
Type: alloc_space
Showing nodes accounting for 37059.20kB, 96.79% of 38287.35kB total
Dropped 1209 nodes (cum <= 191.44kB)
      flat  flat%   sum%        cum   cum%
29917.87kB 78.14% 78.14% 29917.87kB 78.14%  s1::util::StdNewDeleteResource::Allocate
         ... other items ...
         0     0% 96.79%  9625.52kB 25.14%  TiXmlAttribute::operator new
         0     0% 96.79% 25661.21kB 67.02%  TiXmlDocument::LoadFile
         0     0% 96.79% 25657.21kB 67.01%  TiXmlDocument::Parse
         0     0% 96.79% 27344.76kB 71.42%  TiXmlDocument::allocate
         0     0% 96.79% 26500.98kB 69.22%  TiXmlElement::Parse
         0     0% 96.79% 26063.46kB 68.07%  TiXmlElement::ReadValue
         0     0% 96.79% 17719.24kB 46.28%  TiXmlElement::operator new
         ... other items ...
```

We can see that there were allocations of ~ 38 MB, most of that coming from `TiXml*` classes through `s1::util::StdNewDeleteResource::Allocate`. However, in a later dump `0002.heap`, taken after the `TiXml` classes have released their memory, and again processed with `-alloc_space` option, we get the following output:

```
Type: alloc_space
Showing nodes accounting for 12040.94kB, 95.46% of 12613.90kB total
Dropped 1271 nodes (cum <= 63.07kB)
      flat  flat%   sum%        cum   cum%
 3418.61kB 27.10% 27.10%  3418.61kB 27.10%  s1::util::StdNewDeleteResource::Allocate
         ... other items ...
```

We see the absolute number of memory allocated dropped to ~ 12 MB. And no trace of `TiXml*` classes. How is that possible? Shouldn't `-alloc_space` and `-alloc_objects` numbers be ever-increasing with consecutive dumps, and catch all allocations performed until that point (including -- and this is essential here -- those that were deallocated)?

This appears to be a bug, and makes it impossible for us to use gperftools reliably, because we don't know whether it catches all allocation-deallocation peaks done in between two dumps.

What do you think the cause could be? Thanks much for your help.

Architecture we are using: aarch64

May be related to #1388

**Comment by alk** [comment]
Thanks for raising this. So heap profiler only dumps 1 meg worth of "stuff" each time it wakes up.

See here: https://github.com/gperftools/gperftools/blob/d1a0cbe1bfd8bfaa6223fb9913fa872a760e536e/src/heap-profiler.cc#L253

I don't know if it is "it" or you're facing some other artifact (I'd also double check what pprof does; keep in mind that your can do a closer inspection of profiles by doing pprof --raw (go pprof of course)).

We actually have a DCHECK (i.e. when NDEBUG isn't set) that crashes the process if heap exceeds the limit (!!), here: https://github.com/gperftools/gperftools/blob/d1a0cbe1bfd8bfaa6223fb9913fa872a760e536e/src/heap-profile-table.cc#L384

So you should be able to check if this condition affects your case or it is something else.

Let me know what you think.

**Comment by sangelovic** [comment]
@alk Thanks for your response.

Yes, each and every dump is ~ 1 MB large (with minor differences in the range of bytes). Is that OK? Logically I would expect each consecutive dump to be larger or equal to the previous dump, since many more allocations/deallocations from many new places occur in the mean time...

I don't fully understand your point about kProfilerBufferSize. Shall I try to increase kProfilerBufferSize, compile gperftools and see if it helps?

Regarding RAW_DCHECK, I will try it tomorrow and let you know.

**Comment by alk** [comment]
Let me try to rephrase. We dump top-N of samples that add to this limit of 1 meg. Likely explains slight mismatches you see

**Comment by sangelovic** [comment]
Hi @alk. I've just tried to enable `RAW_DCHECK` by appending `-UNDEBUG` to gperftools build flags, but no crash. I then tried increase `kProfilerBufferSize` and re-build. This helped. I had to increase to 32 Meg, since dumps from our application needed from 5 - 22 Megs. Now the data really seems to be complete and trustworthy, no drops and no illogical decreases in numbers. Thanks!

I was not aware of such an internal detail that so greatly affects the completeness and trustworthiness of data. Did I miss related documentation somewhere? I guess this may be an issue to many other users as well.

Now, we have to patch internal gperftools code in our build system to increase its buffer size. This is very cumbersome. Is it possible to customize it at run time (e.g. through env variables or via the library API)? That would be much better.
